repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java

```diff
@@ -2225,118 +2211,71 @@ private void addEdgesToContext(String guid, List<Object> newElementsCreated, Lis
         }
     }
 
-    private void addInternalProductAttr (AttributeMutationContext ctx, List<Object> allArrayElements, List<AtlasEdge> deletedElements) {
-        MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("addInternalProductAttr");
+    private void addInternalProductAttr(AttributeMutationContext ctx, List<Object> createdElements, List<AtlasEdge> deletedElements, List<Object> currentElements) throws AtlasBaseException {
+        MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("addInternalProductAttrForAppend");
         AtlasVertex toVertex = ctx.getReferringVertex();
         String toVertexType = getTypeName(toVertex);
+        List<String> portGuids = new ArrayList<>();
 
         if (TYPE_PRODUCT.equals(toVertexType)) {
-            List<String> allArrayElementGuids = allArrayElements.stream()
-                    .map(x -> ((AtlasEdge) x).getOutVertex().getProperty("__guid", String.class))
-                    .collect(Collectors.toList());
 
-            // Removing deleted elements from allArrayElementGuids
-            if (CollectionUtils.isNotEmpty(deletedElements)) {
-                List<String> deletedGuids = deletedElements.stream()
-                        .map(x -> x.getOutVertex().getProperty("__guid", String.class))
-                        .collect(Collectors.toList());
-                allArrayElementGuids.removeAll(deletedGuids);
+            if(currentElements == null){
+                AtlasAttribute attribute           = ctx.getAttribute();
+                AtlasArrayType arrType             = (AtlasArrayType) attribute.getAttributeType();
+                AtlasType      elementType         = arrType.getElementType();
+                boolean        isStructType        = (TypeCategory.STRUCT == elementType.getTypeCategory()) ||
+                        (TypeCategory.STRUCT == attribute.getDefinedInType().getTypeCategory());
+                boolean        isReference         = isReference(elementType);
+                boolean        isSoftReference     = ctx.getAttribute().getAttributeDef().isSoftReferenced();
+
+                if (isReference && !isSoftReference) {
+                    currentElements = (List) getCollectionElementsUsingRelationship(ctx.getReferringVertex(), attribute, isStructType);
+                } else {
+                    currentElements = (List) getArrayElementsProperty(elementType, isSoftReference, ctx.getReferringVertex(), ctx.getVertexProperty());
+                }
```

Comment:
isReference && !isSoftReference

Since we are dealing with only a specific relationship here, I think these values will be always same & we can remove this & call required directly